network/data_link/at_parser.c

```diff
@@ -0,0 +1,1158 @@
+/******************************************************************************/
+
+#include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include "at_parser.h"
+#include "error.h"
+#include "circular_buffer.h"
+#include "delay.h"
+#include "uart.h"
+#include "irq.h"
+#include "util.h"
+
+/******************************************************************************/
+/********************** Macros and Constants Definitions **********************/
+/******************************************************************************/
+
+/* Should be sizeof(async_msgs)/sizeof(*async_msgs) */
+#define NB_ASYNC_MESSAGES	1
+/* Should be sizeof(responses)/sizeof(*responses) */
+#define NB_RESPONSE_MESSAGES	3
+/* Max command length: at+cwsap=max_ssid_32,max_pass_64,0,0 -> 110 characters */
+#define CMD_BUFF_LEN		120u
+/* Maybe this could be smaller. Here must one response at a time */
+#define RESULT_BUFF_LEN		500u
+/* Used to remove warnings on strings */
+#define PUI8(X)			((uint8_t *)(X))
+/* Timeout waiting for module response. (20 seconds) */
+#define MODULE_TIMEOUT		20000
+/* Delay needed when executing reset command */
+#define RESET_DELAY		6000
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+/* Command descriptor */
+struct cmd_desc {
+	struct at_buff	cmd;
+	uint32_t	type;
+};
+
+/* Structure that links the command string with the command macro by index */
+static const struct cmd_desc g_map[] = {
+	{{(uint8_t *)"", 0}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+RST", 4}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+GSLP", 5}, AT_SET_OP},
+	{{(uint8_t *)"+GMR", 4}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CWMODE", 7}, AT_QUERY_OP | AT_SET_OP | AT_TEST_OP},
+	{{(uint8_t *)"+CWJAP", 6}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+CWLAP", 6}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CWQAP", 6}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CWSAP", 6}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+CWLIF", 6}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CIPSTATUS", 10}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CIPSTART", 9}, AT_TEST_OP | AT_SET_OP},
+	{{(uint8_t *)"+CIPSEND", 8}, AT_SET_OP},
+	{{(uint8_t *)"+CIPCLOSE", 9}, AT_EXECUTE_OP | AT_SET_OP},
+	{{(uint8_t *)"+CIFSR", 6}, AT_EXECUTE_OP},
+	{{(uint8_t *)"+CIPMUX", 7}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+CIPSERVER", 10}, AT_SET_OP},
+	{{(uint8_t *)"+CIPMODE", 8}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+CIPSTO", 7}, AT_QUERY_OP | AT_SET_OP},
+	{{(uint8_t *)"+PING", 5}, AT_SET_OP}
+};
+
+/* Structure storing a circular buffer for a connection */
+struct at_circular_buff {
+	/* Size of the buffer */
+	uint32_t	size;
+	/* Address to the buffer */
+	uint8_t		*buff;
+	/* Read index */
+	uint32_t	ri;
+	/* Write index */
+	uint32_t	wi;
+	/* Write index */
+	uint32_t	count;
```

Comment:
Maybe the comments can be more specific how the `wi` and `count` differ fr each other, or why do we need them both.

Reply:
Copy paste error :) will fix

Reply:
fixed